File: war2/server/trunk/Common/src/war2/common/io/SyncIoWorkProcedure.java
package war2.common.io;

/**
 * 同步的 IO 工作过程
 * 
 * @author haijiang
 *
 */
class SyncIoWorkProcedure<E extends Enum<E>> implements IIoWorkProcedure<IIoWork, E> {
	@Override
	public void startWork(IIoWork work, E threadEnum) {
		if (work == null) {
			return;
		}

		// 将异步操作包装成一个有状态的对象, 
		// 然后带入 callInit, callAsyncProc, callFinish 等函数中!
		this.nextStep(new StatefulIoWork<E>(work, threadEnum));
	}

	/**
	 * 调用异步操作对象的 doInit 函数
	 * 
	 * @param work
	 */
	private void callInit(StatefulIoWork<E> work) {
		if (work == null) {
			return;
		}
	
		work.doInit();
		this.nextStep(work);
	}

	/**
	 * 调用异步操作对象的 doAsyncProc 函数
	 * 
	 * @param work
	 */
	private void callAsyncProc(StatefulIoWork<E> work) {
		if (work == null) {
			return;
		}

		work.doAsyncProc();
		this.nextStep(work);
	}

	/**
	 * 调用异步操作对象的 doFinish 函数
	 * 
	 * @param work
	 */
	private void callFinish(StatefulIoWork<E> work) {
		if (work == null) {
			return;
		}

		work.doFinish();
	}

	/**
	 * 执行下异步操作
	 * 
	 * @param work
	 */
	private void nextStep(StatefulIoWork<E> work) {
		if (work == null) {
			return;
		}

		// 获取当前工作状态
		IoWorkStateEnum currState = work.getCurrState();
	
		if (currState == null) {
			this.callInit(work);
			return;
		}
	
		switch (work.getCurrState()) {
		case initOk:
			// 执行异步过程
			this.callAsyncProc(work);
			return;
	
		case asyncProcOk:
			// 执行结束过程
			this.callFinish(work);
			return;
	
		default:
			return;
		}
	}
}

File: server/Common/src/xgame/common/io/impl/SyncIoWorkProcedure.scala
package xgame.common.io.impl

import xgame.common.io.{TIoWork, TIoWorkProcedure}

/**
 * 同步 IO 工作过程
 *
 * @author haijiang
 * @since 2013/6/17
 *
 */
private[io] class SyncIoWorkProcedure extends TIoWorkProcedure[TIoWork] {
    // @Override
    def startWork(work : TIoWork) = {
        // 如果参数为空, 则直接跳过!
        if (work == null) Unit
        // 直接进入下一步操作
        this.nextStep(new StatefulIoWork(work))
    }

    /**
     * 执行下一步操作
     *
     * @param work
     *
     */
    private[this] def nextStep(work : StatefulIoWork) : Unit = {
        // 如果参数为空, 则直接跳过!
        if (work == null) Unit

        work.currState match {
            // 初始化 IO 工作
            case IoWorkStateEnum.nil => work.doInit()
            // 执行异步过程
            case IoWorkStateEnum.initOk => work.doAsyncProc()
            // 完成 IO 工作
            case IoWorkStateEnum.asyncProcOk => work.doFinished()
            // 其他状态 ...
            case _ =>
        }

        // 继续下一步操作
        this.nextStep(work)
    }
}
